extensions/omni_var/omni_var.c

```diff
@@ -0,0 +1,231 @@
+} VariableValue;
+
+typedef struct {
+  NameData name;
+  VariableValue variable_value;
+} Variable;
+
+static int nvars = 0;
+
+void _PG_init() {
+  DefineCustomIntVariable(
+      "omni_var.estimated_initial_txn_variables_count", "Estimated number of transaction variables",
+      "Estimated number of transaction variables, changes apply to the next transaction", &nvars,
+      1024, 0, 65535, PGC_USERSET, 0, NULL, NULL, NULL);
+}
+
+#if PG_MAJORVERSION_NUM < 14
+#define TXN_HASH_ELEM HASH_ELEM
+#else
+#define TXN_HASH_ELEM HASH_ELEM | HASH_STRINGS
+#endif
+
+static bool subtransaction_callback_registered = false;
+
+static void subtransaction_callback(SubXactEvent event, SubTransactionId xact_id,
+                                    SubTransactionId parent_xact_id, void *arg) {
+  switch (event) {
+  case SUBXACT_EVENT_ABORT_SUB:
+    // Aborting subtransaction xact_id
+
+    if (current_tab == NULL) {
+      // Nothing to process, bail
+      return;
+    }
+    HASH_SEQ_STATUS seq_status;
+    Variable *var;
+
+    hash_seq_init(&seq_status, current_tab);
+    // For every variable:
+    while ((var = (Variable *)hash_seq_search(&seq_status)) != NULL) {
+
+      VariableValue *vvar = &var->variable_value;
+
+      while (vvar != NULL) {
+        if (vvar->subxactid >= xact_id) {
+          // If the value was set at this subxact ID or higher (at or after)
+          if (vvar->previous == NULL) {
+            // No previous values, the whole record is invalid
+            bool found;
+            hash_search(current_tab, NameStr(var->name), HASH_REMOVE, &found);
+            break;
+          } else {
+            // Go up the chain
+            vvar = vvar->previous;
+          }
+        } else {
+          // If it was set at a lower subxact, use it
+          if (vvar != &var->variable_value) {
+            // Copy if it is not the top value.
+            // We are not freeing up other values that may stand between the var and the value
+            // as they will be freed with the transaction.
+            memcpy(&var->variable_value, vvar, sizeof(var->variable_value));
+          }
+          break;
+        }
+      }
+    }
+  default:
+    return;
+  }
+}
+
+static bool transaction_callback_registered = false;
+
+void transaction_callback(XactEvent event, void *arg) {
+  switch (event) {
+  case XACT_EVENT_ABORT:
+  case XACT_EVENT_COMMIT:
+  case XACT_EVENT_PARALLEL_ABORT:
+  case XACT_EVENT_PARALLEL_COMMIT:
+    if (subtransaction_callback_registered) {
+      UnregisterSubXactCallback(subtransaction_callback, NULL);
+      subtransaction_callback_registered = false;
+    }
+    if (transaction_callback_registered) {
+      UnregisterXactCallback(transaction_callback, NULL);
+      transaction_callback_registered = false;
+    }
+  default:
+    // nothing to do
+    return;
+  }
+}
+
+Datum set(PG_FUNCTION_ARGS) {
+  if (PG_ARGISNULL(0)) {
+    ereport(ERROR, errmsg("variable name must not be a null"));
```

Comment:
should we also make sure that it is of type `text` only or does it support other types too?

Reply:
See the documentation and thr Variable type. It's intentionally and expressly typed.